components/script/body.rs

```diff
@@ -234,27 +234,28 @@ impl TransmitBodyConnectHandler {
             task!(setup_native_body_promise_handler: move || {
                 let rooted_stream = stream.root();
                 let global = rooted_stream.global();
+                let cx = GlobalScope::get_cx();
 
                 // Step 4, the result of reading a chunk from body’s stream with reader.
                 let promise = rooted_stream.read_a_chunk(CanGc::note());
 
                 // Step 5, the parallel steps waiting for and handling the result of the read promise,
                 // are a combination of the promise native handler here,
                 // and the corresponding IPC route in `component::net::http_loader`.
-                let promise_handler = Box::new(TransmitBodyPromiseHandler {
+                rooted!(in(*cx) let mut promise_handler = Some(TransmitBodyPromiseHandler {
                     bytes_sender: bytes_sender.clone(),
                     stream: Dom::from_ref(&rooted_stream.clone()),
                     control_sender: control_sender.clone(),
-                });
+                }));
 
-                let rejection_handler = Box::new(TransmitBodyPromiseRejectionHandler {
+                rooted!(in(*cx) let mut rejection_handler = Some(TransmitBodyPromiseRejectionHandler {
                     bytes_sender,
                     stream: Dom::from_ref(&rooted_stream.clone()),
                     control_sender,
-                });
+                }));
 
                 let handler =
-                    PromiseNativeHandler::new(&global, Some(promise_handler), Some(rejection_handler), CanGc::note());
```

Comment:
I did this as suggested in https://github.com/servo/servo/issues/37331 to make it compile, but I think it's wrong. We moved actual value from rooted, we root/trace `None` when doing alloc (and possibly GC) in `PromiseNativeHandler::new`, actual boxed handlers are just floating around freely not being traced.

Reply:
cc @jdm @gterzian

Reply:
The biggest problem is that args (inner fields) need to be rooted (traced) before `reflect_dom_object`. One idea is that we would always make sure that unrooted dom struct is rooted (traced), before creating it's reflector (JSObject).